From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: "Thomas H.P. Andersen" <phomes@gmail.com>
Date: Thu, 15 Oct 2026 18:10:47 +0200
Subject: [PATCH] nvk: fix color_write_enable

I mixed up add/and when I implemented this. Because of this we never
detect when the mask is same for all and always set SET_SINGLE_CT_WRITE_CONTROL
off. The result was still correct but we missed the perf oportunity.

This fixes it and adds mme tests for it

Assisted-by: LLM
Reviewed-by: Mel Henning <mhenning@darkrefraction.com>
Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/45264>
(cherry picked from commit 0f737d4e20d9aabc5164d04c94edf6488db003c5)
---
 src/nouveau/vulkan/nvk_cmd_draw.c | 46 ++++++++++++++++++++++++++++++-
 src/nouveau/vulkan/nvk_mme.c      |  1 +
 src/nouveau/vulkan/nvk_mme.h      |  1 +
 3 files changed, 47 insertions(+), 1 deletion(-)

diff --git a/src/nouveau/vulkan/nvk_cmd_draw.c b/src/nouveau/vulkan/nvk_cmd_draw.c
index 0b3bb8de..01191cc3 100644
--- a/src/nouveau/vulkan/nvk_cmd_draw.c
+++ b/src/nouveau/vulkan/nvk_cmd_draw.c
@@ -4554,7 +4554,7 @@ nvk_mme_set_write_mask(struct mme_builder *b)
       mme_free_reg(b, val);
 
       /* Check if all masks are common */
-      struct mme_value temp = mme_add(b, mask, mme_imm(BITFIELD_RANGE(0, 4)));
+      struct mme_value temp = mme_and(b, mask, mme_imm(BITFIELD_RANGE(0, 4)));
       mme_if(b, ine, first, temp) {
          mme_mov_to(b, common_mask, mme_zero());
       }
@@ -4569,6 +4569,50 @@ nvk_mme_set_write_mask(struct mme_builder *b)
    mme_emit(b, common_mask);
 }
 
+const struct nvk_mme_test_case nvk_mme_set_write_mask_tests[] = {{
+   .params = (uint32_t[]) { 1, 0x0 },
+   .expected = (struct nvk_mme_mthd_data[]) {
+      { NV9097_SET_CT_WRITE(0), 0x0000 },
+      { NV9097_SET_SINGLE_CT_WRITE_CONTROL, 1 },
+      { }
+   },
+}, {
+   .params = (uint32_t[]) { 1, 0xf },
+   .expected = (struct nvk_mme_mthd_data[]) {
+      { NV9097_SET_CT_WRITE(0), 0x1111 },
+      { NV9097_SET_SINGLE_CT_WRITE_CONTROL, 1 },
+      { }
+   },
+}, {
+   .params = (uint32_t[]) { 8, 0x0f0f8421 },
+   .expected = (struct nvk_mme_mthd_data[]) {
+      { NV9097_SET_CT_WRITE(0), 0x0001 },
+      { NV9097_SET_CT_WRITE(1), 0x0010 },
+      { NV9097_SET_CT_WRITE(2), 0x0100 },
+      { NV9097_SET_CT_WRITE(3), 0x1000 },
+      { NV9097_SET_CT_WRITE(4), 0x1111 },
+      { NV9097_SET_CT_WRITE(5), 0x0000 },
+      { NV9097_SET_CT_WRITE(6), 0x1111 },
+      { NV9097_SET_CT_WRITE(7), 0x0000 },
+      { NV9097_SET_SINGLE_CT_WRITE_CONTROL, 0 },
+      { }
+   },
+}, {
+   .params = (uint32_t[]) { 8, 0xffffffff },
+   .expected = (struct nvk_mme_mthd_data[]) {
+      { NV9097_SET_CT_WRITE(0), 0x1111 },
+      { NV9097_SET_CT_WRITE(1), 0x1111 },
+      { NV9097_SET_CT_WRITE(2), 0x1111 },
+      { NV9097_SET_CT_WRITE(3), 0x1111 },
+      { NV9097_SET_CT_WRITE(4), 0x1111 },
+      { NV9097_SET_CT_WRITE(5), 0x1111 },
+      { NV9097_SET_CT_WRITE(6), 0x1111 },
+      { NV9097_SET_CT_WRITE(7), 0x1111 },
+      { NV9097_SET_SINGLE_CT_WRITE_CONTROL, 1 },
+      { }
+   },
+}, {}};
+
 static void
 nvk_flush_cb_state(struct nvk_cmd_buffer *cmd)
 {
diff --git a/src/nouveau/vulkan/nvk_mme.c b/src/nouveau/vulkan/nvk_mme.c
index 4441c676..ef041c76 100644
--- a/src/nouveau/vulkan/nvk_mme.c
+++ b/src/nouveau/vulkan/nvk_mme.c
@@ -53,6 +53,7 @@ static const struct nvk_mme_test_case *mme_tests[NVK_MME_COUNT] = {
    [NVK_MME_SET_SHADING_RATE_CONTROL]      = nvk_mme_set_shading_rate_control_tests,
    [NVK_MME_SET_ANTI_ALIAS]                = nvk_mme_set_anti_alias_tests,
    [NVK_MME_SET_STATISTICS_COUNTERS]       = nvk_mme_set_statistics_counters_tests,
+   [NVK_MME_SET_WRITE_MASK]                = nvk_mme_set_write_mask_tests,
 };
 
 uint32_t *
diff --git a/src/nouveau/vulkan/nvk_mme.h b/src/nouveau/vulkan/nvk_mme.h
index 12b11dcd..f95ae71e 100644
--- a/src/nouveau/vulkan/nvk_mme.h
+++ b/src/nouveau/vulkan/nvk_mme.h
@@ -303,6 +303,7 @@ extern const struct nvk_mme_test_case nvk_mme_set_tess_params_tests[];
 extern const struct nvk_mme_test_case nvk_mme_set_shading_rate_control_tests[];
 extern const struct nvk_mme_test_case nvk_mme_set_anti_alias_tests[];
 extern const struct nvk_mme_test_case nvk_mme_set_statistics_counters_tests[];
+extern const struct nvk_mme_test_case nvk_mme_set_write_mask_tests[];
 
 void nvk_test_all_mmes(const struct nv_device_info *devinfo);
 
-- 
2.43.0

